plugins/carlabase/carla.cpp

```diff
@@ -286,14 +311,97 @@ void CarlaInstrument::saveSettings(QDomDocument& doc, QDomElement& parent)
 
     QDomDocument carlaDoc("carla");
-
     if (carlaDoc.setContent(QString(state)))
     {
         QDomNode n = doc.importNode(carlaDoc.documentElement(), true);
         parent.appendChild(n);
     }
-
     std::free(state);
+
+    for (uint32_t i=0; i < floatModels.count(); ++i)
+    {
+        QString idStr = CARLA_SETTING_PREFIX + QString::number(i);
+        floatModels[i]->saveSettings(doc, parent, idStr);
+    }
+}
+
+void CarlaInstrument::refreshParams(bool valuesOnly = false, bool init = false)
+{
+	if (fDescriptor->get_parameter_count != nullptr &&
+		fDescriptor->get_parameter_info  != nullptr &&
+		fDescriptor->get_parameter_value != nullptr &&
+		fDescriptor->set_parameter_value != nullptr)
+	{
+		uint32_t param_count = fDescriptor->get_parameter_count(fHandle);
+
+		if (!valuesOnly)
+		{
+			clearKnobModels();
+			floatModels.reserve(param_count);
+		}
+
+		QList<QString> completerData;
+
+		for (uint32_t i=0; i < param_count; ++i)
+		{
+			// https://github.com/falkTX/Carla/tree/master/source/native-plugins source/native-plugins/resources/carla-plugin
+			float param_value = fDescriptor->get_parameter_value(fHandle, i);
+
+			if (valuesOnly)
+			{
+				floatModels[i]->setValue(param_value);
+				continue;
+			}
+
+			const NativeParameter* paramInfo(fDescriptor->get_parameter_info(fHandle, i));
+
+			// Get parameter name
+			QString name = "_NO_NAME_";
+			if (paramInfo->name != nullptr){
+				name = paramInfo->name;
+			}
+
+			completerData.push_back(name);
+
+			// current_value, min, max, steps
+			floatModels.push_back(new FloatModel(param_value, paramInfo->ranges.min,
+					paramInfo->ranges.max, paramInfo->ranges.step, this, name));
+
+			// Load settings into model.
+			QString idStr = CARLA_SETTING_PREFIX + QString::number(i);
+			floatModels[i]->setObjectName(QString::number(i)); // use this as reference for parameter index.
+			if (init)
+			{
+				floatModels[i]->loadSettings(settingsElem, idStr);
+			}
+
+			// Connect to signal dataChanged to knobChanged function.
+			//connect(floatModels[i], SIGNAL(dataChanged()), this, SLOT(knobModelChanged()));
+			connect(floatModels[i], &FloatModel::dataChanged, [=]() {knobModelChanged(i);});
+		}
+		// Set completer data
+		m_completerModel->setStringList(completerData);
+	}
+}
+
+void CarlaInstrument::clearKnobModels(){
+	//Delete the models, this also disconnects all connections (automation and controller connections)
+	for (uint32_t i=0; i < floatModels.count(); ++i)
+	{
+		delete floatModels[i];
+	}
+
+	//Clear the list
+	floatModels.clear();
+}
+
+void CarlaInstrument::knobModelChanged(uint32_t index)
+{
+	//FloatModel *senderFloatModel = qobject_cast<FloatModel *>(sender());
+	if (fDescriptor->set_parameter_value != nullptr){
+		//fDescriptor->set_parameter_value(fHandle, senderFloatModel->objectName().toInt(), senderFloatModel->value());
+		fDescriptor->set_parameter_value(fHandle, floatModels[index]->objectName().toInt(), floatModels[index]->value());
```

Comment:
I believe `floatModels[index]->objectName().toInt()` can be changed to `index`. Is it correct?

Reply:
Right again :+1: